server/api/src/rsDataObjRepl.cpp

```diff
@@ -477,62 +481,88 @@ namespace
+                return false;
+        }
+    } // permission_is_sufficient_to_replicate_object
+
     auto user_has_permission_to_replicate_object(RsComm& _comm, const irods::file_object_ptr _obj) -> bool
     {
+        namespace adm = irods::experimental::administration;
+
         // Short-circuit the query by checking to see whether this is a rodsadmin using the ADMIN_KW, which allows
         // them to do whatever they want.
         if (irods::is_privileged_client(_comm) && getValByKey(&_obj->cond_input(), ADMIN_KW)) {
             return true;
         }
 
+        const auto user = adm::user{_comm.clientUser.userName, _comm.clientUser.rodsZone};
         const auto path = irods::experimental::filesystem::path{_obj->logical_path()};
-        const auto permission_query_string =
-            fmt::format("select DATA_ACCESS_TYPE where USER_NAME = '{}' and USER_ZONE = '{}' and DATA_ID = '{}'",
-                        _comm.clientUser.userName,
-                        _comm.clientUser.rodsZone,
-                        _obj->data_id());
-
-        auto permission_query = irods::query{&_comm, permission_query_string};
-
-        // If the user has no permissions on the object, it may have returned an empty result set.
-        if (permission_query.empty()) {
-            return false;
-        }
-
-        const auto& query_result = permission_query.front();
-        const auto& access_type_string = query_result[0];
 
         try {
-            using access_type = irods::experimental::catalog::access_type;
-
-            switch (std::stoi(access_type_string)) {
-                case static_cast<int>(access_type::modify_object): // NOLINT(bugprone-branch-clone)
-                    [[fallthrough]];
-                case static_cast<int>(access_type::delete_object): // NOLINT(bugprone-branch-clone)
-                    [[fallthrough]];
-                case static_cast<int>(access_type::own):
-                    return true;
-
-                default:
-                    return false;
+            // Check access permission for the user first...
+            const auto user_permission_query_string =
+                fmt::format("select DATA_ACCESS_TYPE where USER_NAME = '{}' and USER_ZONE = '{}' and DATA_ID = '{}'",
+                            user.name,
+                            user.zone,
+                            _obj->data_id());
+            auto user_permission_query = irods::query{&_comm, user_permission_query_string};
+            if (!user_permission_query.empty() &&
+                permission_is_sufficient_to_replicate_object(std::stoi(user_permission_query.front()[0])))
+            {
+                return true;
             }
+
+            // If the user does not have sufficient permissions on the data object for replication, perhaps one of their
+            // groups does...
+            const auto groups = adm::server::groups(_comm, user);
+            std::vector<std::string> groups_with_quotes;
```

Comment:
Add line to reserve the size of the vector containing quoted group names.
```cpp
groups_with_quotes.reserve(groups.size());
```

Reply:
Sure thing.

By the way, I tried to make some `libfmt` and `std::vector` magic happen here so it would be a bit cleaner but this is the best I could come up with for now. :)

Reply:
Done